platform/mbed_retarget.h

```diff
@@ -94,6 +94,7 @@ namespace mbed {
 class FileHandle;
 class DirHandle;
 
+#if !(MBED_CONF_PLATFORM_STDIO_MINIMAL_CONSOLE_ONLY)
```

Comment:
That highlights that `Stream` won't work if minimal console is enabled - it relies on all the file handle stuff.

Which in turns highlights why I want to deprecate `Stream` and `Serial` which relies on it - it pulls in all that stuff.

As we're doing the serial restructuring anyway, one possible thing to do here would be to make `Stream` (and any classes that rely on it, like `Serial`) conditional on the minimal retarget being off. `UARTSerial`/`BufferedSerial` and `RawSerial`/`UnbufferedSerial` (and `SerialBase`) will work fine.

Reply:
The following USB classes also use `Stream`: `USBKeyboard`, `USBMouseKeyboard`, and `USBSerial`. They will need to be made conditional on the minimal re-target being off along with `Stream` and `Stream`. (Option 1)

Alternatively, only the `Stream` methods that use the re-target code (`Stream::v\printf`, `Stream::v\scanf`) could be removed. (Option 2)

@evedon should we go for option 1 or 2?

Reply:
I don't think it helps to go with option 2. The issue is that `Stream` won't work at all. In the constructor there is a call to `fdopen` which will return NULL. Currently this will result in an MBED_ERROR with `Stream obj failure`.
We could make a check on whether minimal re-target is enabled and provide a more specific message.

Reply:
Well, you'd remove the `fdopen` and `Stream::_file` too, because it wouldn't be needed once you'd removed the methods - that's why you're removing the methods.

I don't feel strongly about the choice between the two options. If 2 is straightforward, I'm fine with it, but wouldn't want to work too hard on it.

There's a remaining paradox in that once you kill all the C-based methods like `printf` and `putc`, all you're left with is the `FileHandle`-overridden `read` and `write`, but they're oddly marked `protected` by Stream. You can't get at them directly like this:

    Serial s;
    s.write(...);

You'd have to do

    Serial s;
    FileHandle &f = s;
    f.write(...);

or

    static_cast<FileHandle &>(s).write(...);

Maybe those methods could be flipped to public. They are effectively public anyway, because they're overrides of public methods.

I never made that change because I've never made any change to `Stream` because I think we  should be killing it with fire, not fixing it...

Reply:
If you really wanted to keep stuff limping on, you could provide alternative definitions of some of the `Stream` methods for minimal console that avoid the C library.

Eg `Stream::putc` could become a call to `Stream::_putc(c)` rather than to `::fputc(c, _file)`.

But I fear that might just add to the confusion.

Reply:
This [force-push](https://github.com/ARMmbed/mbed-os/compare/d3aa3b461ee7d9f998560072a0c608d0d22ccd43..3bbd53d11003c1862ff2b3918b460435e591358c) excludes all references of `_file` in `Stream` and make `Stream::read` and `Stream::write` public.

Reply:
That's a nice solution.